README
hgtools
=======

hgtools builds on the setuptools_hg plugin for setuptools. hgtools
provides classes for inspecting and working with repositories in the
Mercurial version control system.

hgtools provides a plugin for setuptools that enables setuptools to find
files under the Mercurial version control system.

The classes provided by hgtools are designed to work natively with the
Mercurial Python libraries (in process) or fall back to using the
command-line program ``hg(1)`` if available. The command-line support is
especially useful inside virtualenvs
that don't have access to a system-wide installed Mercurial lib (i.e. when
the virtualenv was created with ``--no-site-packages``).

.. note:: The setuptools feature

  You can read about the setuptools plugin provided by hgtools in the
  `setuptools documentation`_. It basically returns a list of files that are
  under Mercurial version control when running the ``setup`` function, e.g. if
  you create a source and binary distribution. It's a simple yet effective way
  of not having to define package data (non-Python files) manually in MANIFEST
  templates (``MANIFEST.in``).

.. _setuptools documentation: http://peak.telecommunity.com/DevCenter/setuptools#adding-support-for-other-revision-control-systems

Usage
*****

Here's a simple example of a setup.py that uses hgtools::

    from setuptools import setup, find_packages
    setup(
        name="HelloWorld",
        version="0.1",
        packages=find_packages(),
        setup_requires=["hgtools"],
    )

If you run the setup.py above, setuptools will automatically download
hgtools to the directory where the setup.py is located at (and won't
install it anywhere else) to get all package data files from the
Mercurial repository.

You should not need to, and I recommend you don't, install hgtools in
your site-packages directory. Let setuptools grab it on demand. Also,
try not to specify an upper bound for the requirement. Usually, simply
specifying 'hgtools' will get the latest version, which is likely to
remain compatible (as a plugin) for the life of the project. Specifying
an upper bound (i.e. `hgtools<1.1`) will only prevent you from getting
bug fixes. Only specify an upper bound if you require support for older
versions of Python.

Auto Version Numbering
**********************

With the 0.4 release, hgtools adds support for automatically generating
project version numbers from the mercurial repository in which the
project is developed.

To use this feature, your project must follow the following assumptions:

	 - Mercurial tags are used to indicate released versions.
	 - Tag names are specified as the version only (i.e. 0.1 and not
	   v0.1 or release-0.1)
	 - Released versions currently must conform to the StrictVersion in
	   distutils. Any tags that don't match this scheme will be ignored.
	   Future releases may relax this restriction.

Thereafter, you may use the HGToolsManager.get_current_version to
determine the version of your product. If the current revision is tagged
with a valid version, that version will be used. Otherwise, the tags in
the repo will be searched, the latest release will be found, and hgtools
will infer the upcoming release version.

For example, if the repo contains the tags 0.1, 0.2, and 0.3 and the
repo is not on any of those tags, get_current_version will return
'0.3.1dev' and get_current_version(increment='0.1') will return
'0.4dev'.

A distutils hook has been created to hack setuptools to use this version
information automatically. To use this functionality, just use the
``use_hg_version`` parameter to setup.
For example::

    from setuptools import setup, find_packages
    setup(
        name="HelloWorld",
        use_hg_version=True,
        packages=find_packages(),
        setup_requires=["hgtools"],
    )

If the value supplied to use_hg_version resolves to True, hgtools will
use the mercurial version to determine the version of the
package (based on get_current_version). If an sdist is created, hgtools
will store the calculated version in the tag_build of the setup.cfg and
will use that version when deploying remotely. Therefore, if you are
using auto-versioning, you should not use setuptools tags explicitly.

See the jaraco.util setup.py for an example of this technique.

Versioning Parameters
~~~~~~~~~~~~~~~~~~~~~

It's also possible to pass keyword parameters to use_hg_version to
tweak how it generates version numbers. To pass parameters, instead of
setting `use_hg_version = True`, set it to a non-empty dictionary with
one or more of the following parameters:

 - `increment`:
   A string indicating the default version increment for the project.
   By default, this value is '0.1', meaning hgtools will use the version
   '1.1dev' for builds following the 1.0 release and '1.10dev' for builds
   following a 1.9.3 release. Set this value to '1.0' or '0.0.1' for the
   current tree to help hgtools guess the target version.

 - `version_handler`:
   A Python function with the following signature::

       def calc_version(mgr, options):
           return str('1.0')

   hgtools will use this function instead of its default implementation
   to customize the version number calculation. The `mgr` object is the
   `hgtools.managers.HGRepoManager` object referencing the local repo
   and the `options` is the dictionary passed to use_hg_version.

   Use this option, for example, to include the mercurial hash or local
   revision ID in the version::

       def id_as_version(mgr, options):
           "Always return the mercurial revision ID as the version"
           id_n = mgr._run_cmd([mgr.exe, 'id', '-n']).strip()
           return id_n

       setup(
           #...
           use_hg_version={'version_handler': id_as_version},
       )

   The first thing to note is the mgr does not yet provide a nice
   interface for getting anything but the tags for a revision, so the
   example digs into the underlying API to extract the ID. hgtools should
   provide better support in the HGRepoManager classes in future releases.

   Use this feature with caution. If you have not already read the
   `setuptools documentation on specifying a project version
   <http://packages.python.org/distribute/setuptools.html#specifying-your-project-s-version>`_,
   the author recommends you do read that.


CHANGES
Changes
*******

4.0
~~~

* Removed functions in hgtools (package) module deprecated since 0.6.6.

3.0.2
~~~~~

* Restored Python 2.6 compatibility in reentry module.

3.0.1
~~~~~

* Merged changes from 2.2.2 to address #13.

3.0
~~~

* Removed support for Python 2.5. Python 2.6 or later is now required.

2.2.3
~~~~~

* Restored Python 2.5 compatibility in some modules.

2.2.2
~~~~~

* Removed the LibraryManager from the default imports. This means it will not
  be available unless the client application specifically imports
  `hgtools.managers.library`. This change appears to fix #13.

2.2.1
~~~~~

* Added some exception handling and tests around Library Manager in the hopes
  of preventing #13.

2.2
~~~

* LibraryManager has been re-written to use the command-line API in-process.
  The new LibraryManager now only works Python 2.7 and greater, but also now
  shares the same capability as the SubprocessManager with full tags support.

2.1
~~~

* hgtools now builds natively on Python 3, rather than requiring a 2to3
  conversion at install time.

2.0.3
~~~~~

 * Issue #12: Suppress exceptions besides ImportError when importing
   Mercurial modules.

2.0.2
~~~~~

* Force `hg log` and `hg parents` to use the defaults style for output.
  Otherwise, the parsing code fails to identify the tags. Also, reset the
  'default.log' value in case a user has an alias for `hg log` in his .hgrc.
  (For example, some use log=-G to force the graph output). Thanks to
  `dc <https://bitbucket.org/dc>`_ for the patch.

2.0.1
~~~~~

* Fixed issue #10 where hgtools would not parse setup.cfg properly if
  the Python 3 configparser backport was installed on Python 2.

2.0
~~~

* Refactored HGRepoManager to better support multiple tags:

  - `.get_tag` replaced by `.get_tags`, which returns a set of tags
    for a specific revision. This is currently a set because mercurial
    does not retain any meaningful order of the tags.
  - `.get_tags` replaced by `.get_repo_tags`.
  - `.get_parent_tag` replaced by `.get_parent_tags`.
  - added `.get_parents` which returns the revision(s) of the specified
    revision.

* Removed support for older versions of mercurial (LegacyLibraryManager).
* The subprocess manager is now the default. The HGTOOLS_FORCE_CMD variable
  no longer has any effect on hgtools.
* Version detection now resolves multiple tags on the same revision by
  choosing the greatest version.

1.2.1
~~~~~

* Fix issue #9 - The repo managers and thus the setuptools plugin will no
  longer find files that aren't in the location specified. The
  LibraryManagers already will throw an error in this case, but now the
  SubprocessManager does what's best and only returns files relative
  to the location.

1.2
~~~

* Implemented the `version_handler` version parameter. Fixes #5.
* If multiple tags are given for a revision, the last tag is used instead
  of the first.

1.1.6
~~~~~

* More aggressively construct a the environment when running `hg` in a
  subprocess. Fixes another manifestation of #7. Thanks whit537.

1.1.5
~~~~~

* Fix issue #8 - Re-wrote SubprocessManager.get_tag to extract the tag using
  `hg tags` for more reliable tag resolution.

1.1.3
~~~~~

* Fix issue #7 - SubprocessManager now passes explicit environment to child
  process.

1.1.2
~~~~~

* Restored Python 2.5 compatibility.

1.1
~~~

* Added support for subrepos. The setuptools plugin will now traverse
  subrepos when finding files.

1.0.1
~~~~~

* Fix issue #6 where the force_hg_version flag was affecting installation
  of tagged packages not employing hgtools.

1.0
~~~

* Python 3 support
* Now supports revisions with multiple tags (chooses the first, which
  appears to be the latest).
* Removed support for deprecated use_hg_version_increment.
* Added HGRepoManager.existing_only to filter managers for only those
  which refer to an existing repo.
* Employed HGRepoManager.existing_only in plugins. Fixes #2.
* SubprocessManager no longer writes to /dev/null. Fixes #3.

0.6.7
~~~~~

* Auto-versioning will no longer use the parent tag if the working
  copy has modifications.

0.6.6
~~~~~

* Some minor refactoring - moved functions out of top-level `hgtools`
  module into hgtools.plugins.

0.6.5
~~~~~
 * Test case and fix for error in SubprocessManager when 'hg'
   executable doesn't exist.

0.6.4
~~~~~
 * Fix for NameError created in 0.6.3.

0.6.3
~~~~~
 * Deprecated use_hg_version_increment setup parameter in favor of
   parameters to use_hg_version.

0.6.2
~~~~~
 * From drakonen: hgtools will now utilize the parent changeset tag
   for repositories that were just tagged (no need to update to that
   tag to release).

0.6.1
~~~~~
 * Fixed issue #4: Tag-based autoversioning fails if hgrc defaults
   used for hg identify

0.6
~~~
 * Refactored modules. Created ``managers``, ``versioning``, and
   ``py25compat`` modules.

0.5.2
~~~~~
 * Yet another fix for #1. It appears that simply not activating the
   function is not sufficient. It may be activated by previously-
   installed packages, so it needs to be robust for non-hgtools
   packages.

0.5.1
~~~~~
 * Fix for issue #1 - version_calc_plugin is activated for projects that
   never called for it.
 * LibraryManagers no longer raise errors during the import step
   (instead, they just report as being invalid).
 * SubprocessManager now raises a RuntimeError if the executed command
   does not complete with a success code.

0.5
~~~
 * Fixed issue in file_finder_plugin where searching for an
   appropriate manager would fail if mercurial was not installed in
   the Python instance (ImportErrors weren't trapped properly).

0.4.9
~~~~~
 * Fixed issue where version calculation would fail if tags contained
   spaces.

0.4.8
~~~~~
 * Auto versioning now provides a reasonable default when no version
   tags are yet present.

0.4.3-0.4.7
~~~~~~~~~~~
 * Fixes for versions handling of hgtools itself.

0.4.2
~~~~~
 * Fixed formatting errors in documentation.

0.4.1
~~~~~


hgtools.py
"""
Classes for working with repositories using the Mercurial version
control system. Includes a plugin for setuptools to find files managed
by the version control system.

Classes support using the native Python library interfaces or using the
hg(1) command as a subprocess.
"""
__version__ = '0.4'
__author__ = 'Jannis Leidel/Jason R. Coombs'
__all__ = ['file_finder_plugin', 'get_manager', 'HGRepoManager']

import os
import subprocess
try:
	from collections import namedtuple
except ImportError:
	# Python 2.5 compat
	from namedtuple_backport import namedtuple
from distutils.version import StrictVersion
import operator

def find(pred, items):
	"""
	Find the index of the first element in items for which pred returns
	True
	>>> find(lambda x: x > 3, range(100))
	4
	>>> find(lambda x: x < -3, range(100)) is None
	True
	"""
	for i, item in enumerate(items):
		if pred(item): return i

def rfind(pred, items):
	"""
	Find the index of the last element in items for which pred returns
	True. Returns a negative number useful for indexing from the end
	of a list or tuple.
	
	>>> rfind(lambda x: x > 3, [5,4,3,2,1])
	-4
	"""
	return -find(pred, reversed(items))-1

class SummableVersion(StrictVersion):
	"""
	A special version of a StrictVersion which can be added to another
	StrictVersion.
	
	>>> SummableVersion('1.1') + StrictVersion('2.3')
	SummableVersion ('3.4')
	"""
	def __add__(self, other):
		result = SummableVersion(str(self))
		result.version = tuple(map(operator.add, self.version, other.version))
		return result

	def reset_less_significant(self, significant_version):
		"""
		Reset to zero all version info less significant than the
		indicated version.
		>>> ver = SummableVersion('3.1.2')
		>>> ver.reset_less_significant(SummableVersion('0.1'))
		>>> str(ver)
		'3.1'
		"""
		nonzero = lambda x: x != 0
		version_len = 3 # strict versions are always a tuple of 3
		significant_pos = rfind(nonzero, significant_version.version)
		significant_pos = version_len + significant_pos + 1
		self.version = (self.version[:significant_pos]
			+ (0,)*(version_len - significant_pos) )

	def as_number(self):
		"""
		>>> str(SummableVersion('1.9.3').as_number())
		'1.93'
		"""
		def combine(subver, ver):
			return subver/10.0 + ver
		return reduce(combine, reversed(self.version))

class VersionManagement(object):
	"""
	Version functions for HGRepoManager classes
	"""
	
	increment = '0.0.1'

	def get_strict_versions(self):
		"""
		Return all version tags that can be represented by a
		StrictVersion.
		"""
		for tag in self.get_tags():
			try:
				yield StrictVersion(tag.tag)
			except ValueError:
				pass

	def get_tagged_version(self):
		"""
		Get the version of the local repository as a StrictVersion or
		None if no viable tag exists.
		"""
		try:
			# use 'xxx' because StrictVersion(None) is apparently ok
			return StrictVersion(self.get_tag() or 'xxx')
		except ValueError:
			pass

	def get_latest_version(self):
		"""
		Determine the latest version ever released of the project in
		the repo (based on tags).
		"""
		versions = sorted(self.get_strict_versions(), reverse=True)
		next = lambda i: i.next()
		return next(iter(versions))

	def get_current_version(self, increment=None):
		"""
		Return as a string the version of the current state of the
		repository -- a tagged version, if present, or the next version
		based on prior tagged releases.
		"""
		ver = self.get_tagged_version() or str(self.get_next_version(increment))+'dev'
		return str(ver)

	def get_next_version(self, increment=None):
		"""
		Return the next version based on prior tagged releases.
		"""
		increment = increment or self.increment
		return self.infer_next_version(self.get_latest_version(), increment)

	@staticmethod
	def infer_next_version(last_version, increment):
		"""
		Given a simple application version (as a StrictVersion),
		and an increment (1.0, 0.1, or 0.0.1), guess the next version.
		
		# set up a shorthand for examples
		>>> VM_infer = lambda *params: str(VersionManagement.infer_next_version(*params))
		
		>>> VM_infer('3.2', '0.0.1')
		'3.2.1'
		>>> VM_infer(StrictVersion('3.2'), '0.0.1')
		'3.2.1'
		>>> VM_infer('3.2.3', '0.1')
		'3.3'
		>>> VM_infer('3.1.2', '1.0')
		'4.0'
		
		Subversions never increment parent versions
		>>> VM_infer('3.0.9', '0.0.1')
		'3.0.10'
		
		If it's a prerelease version, just remove the prerelease.
		>>> VM_infer('3.1a1', '0.0.1')
		'3.1'
		"""
		last_version = SummableVersion(str(last_version))
		if last_version.prerelease:
			last_version.prerelease = None
			return str(last_version)
		increment = SummableVersion(increment)
		sum = last_version + increment
		sum.reset_less_significant(increment)
		return sum


class HGRepoManager(VersionManagement, object):
	"""
	An abstract class defining some interfaces for working with
	Mercurial repositories.
	"""
	def __init__(self, location='.'):
		self.location = location
		self.setup()

	def is_valid(self):
		return True

	def setup(self):
		pass

	@staticmethod
	def get_valid_managers(location):
		classes = (LibraryManager, LegacyLibraryManager, SubprocessManager)
		force_cmd = os.environ.get('HGTOOLS_FORCE_CMD', False)
		if force_cmd:
			classes = (SubprocessManager, LibraryManager, LegacyLibraryManager)
		managers = (cls(location) for cls in classes)
		return (mgr for mgr in managers if mgr.is_valid())

	@staticmethod
	def get_first_valid_manager(location='.'):
		next = lambda i: i.next()
		return next(HGRepoManager.get_valid_managers(location))

	def __repr__(self):
		class_name = self.__class__.__name__
		loc = self.location
		return '%(class_name)s(%(loc)r)' % vars()

	def find_files(self):
		raise NotImplementedError()

	def get_tag(self):
		raise NotImplementedError()

	def get_tags(self):
		raise NotImplementedError()

class SubprocessManager(HGRepoManager):
	"""
	An HGRepoManager implemented by calling into the 'hg' command-line
	as a subprocess.
	"""
	exe = 'hg'

	def is_valid(self):
		return 0 == subprocess.call(
			[self.exe, 'version'],
			stdout=self._get_devnull(),
			)

	def _run_cmd(self, cmd):
		proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
			cwd=self.location)
		stdout, stderr = proc.communicate()
		return stdout

	def find_files(self):
		"""
		Find versioned files in self.location
		"""
		return self._run_cmd([self.exe, 'locate']).splitlines()

	@staticmethod
	def _get_devnull():
		return open(os.path.devnull, 'w')

	def get_tag(self):
		return self._run_cmd([self.exe, 'identify', '-t']).strip() or None

	def get_tags(self):
		tagged_revision = namedtuple('tagged_revision', 'tag revision')
		lines = self._run_cmd([self.exe, 'tags']).splitlines()
		return (tagged_revision(*line.split()) for line in lines if line)			

class LibraryManager(HGRepoManager):
	"""
	An HGRepoManager implemented by exercising the mercurial Python APIs
	directly.
	"""
	OLD_VERSIONS = ('1.0', '1.0.1', '1.0.2')

	def setup(self):
		try:
			from mercurial.__version__ import version
			from mercurial import hg, ui, cmdutil
		except ImportError:
			pass

		try:
			from mercurial.error import RepoError
		except ImportError:
			# mercurial < 1.2
			from mercurial.repo import RepoError
		del self
		globals().update(vars())

	def is_valid(self):
		return 'hg' in globals() and self.version_match()

	def version_match(self):
		return version not in self.OLD_VERSIONS

	def _get_repo(self):
		return hg.repository(ui.ui(), path=self.location)

	@property
	def repo(self):
		if not hasattr(self, '_repo'):
			self._repo = self._get_repo()
		return self._repo

	def _get_excluded(self):
		"""
		Return all files that hg knows about, but haven't been added,
		deleted, or removed or have an unknown status.
		"""
		modified, added, removed, deleted, unknown = self.repo.status()[:5]
		return removed + deleted + unknown

	def find_files(self):
		"""
		Use the Mercurial library to recursively find versioned files in dirname.
		"""
		excluded = self._get_excluded()
		rev = None
		match = cmdutil.match(self.repo, [], {}, default='relglob')
		match.bad = lambda x, y: False
		return (abs
			for abs in self.repo[rev].walk(match)
			if (rev or abs in self.repo.dirstate)
			and abs not in excluded
			)

class LegacyLibraryManager(LibraryManager):
	"""
	A special subclass of LibraryManager which works with older versions
	of the Mercurial libraries.
	"""
	def version_match(self):
		return version in self.OLD_VERSIONS

	def find_files(self):
		excluded = self._get_excluded()
		from mercurial import util
		node = None
		walker = cmdutil.walk(self.repo, [], {}, node=node,
			badmatch=util.always, default='relglob')
		return (abs
			for src, abs, rel, exact in walker
			if src != 'b' and (node or abs in repo.dirstate)
			and abs not in excluded
			)

def file_finder_plugin(dirname="."):
	"""
	Find the files in ``dirname`` under Mercurial version control
	according to the setuptools spec (see
	http://peak.telecommunity.com/DevCenter/setuptools#adding-support-for-other-revision-control-systems
	).
	"""
	import distutils.log
	dirname = dirname or '.'
	try:
		for mgr in HGRepoManager.get_valid_managers(dirname):
			try:
				return mgr.find_files()
			except BaseException, e:
				distutils.log.warn("Error in hgtools.%s: %s", mgr, e)
	except BaseException, e:
		distutils.log.warn("Error getting managers in hgtools.file_finder_plugin: %s", e)
	return []

def patch_egg_info():
	from setuptools.command.egg_info import egg_info
	from pkg_resources import safe_version
	import functools
	orig_ver = egg_info.tagged_version
	@functools.wraps(orig_ver)
	def tagged_version(self):
		if self.distribution.use_hg_version:
			result = safe_version(self.distribution.get_version())
		else:
			result = orig_ver(self)
		self.tag_build = result
		return result
	egg_info.tagged_version = tagged_version

def version_calc_plugin(dist, attr, value):
	"""
	Handler for parameter to setup(use_hg_version=value)
	"""
	if not value: return
	# we cache the version in the tag_build value in setup.cfg (so that
	#  sdist versions will have a copy of the version as determined at
	#  the build environment).
	from ConfigParser import ConfigParser
	parser = ConfigParser()
	parser.read('setup.cfg')
	has_tag_build = (parser.has_section('egg_info')
		and 'tag_build' in parser.options('egg_info'))
	if has_tag_build:
		# a cached version is available, so use it.
		version = parser.get('egg_info', 'tag_build')
	else:
		# for now, force the CMD version, because the library version
		#  is not implemented.
		os.environ['HGTOOLS_FORCE_CMD'] = 'True'
		mgr = HGRepoManager.get_first_valid_manager()
		# if the user indicates an increment, use it
		increment = value if 'increment' in attr else None
		version = mgr.get_current_version(increment)
	dist.metadata.version = version
	patch_egg_info()


# kept for backward-compatibility
get_manager = HGRepoManager.get_first_valid_manager
